x-pack/legacy/plugins/lens/public/datatable_visualization/visualization.tsx

```diff
@@ -188,17 +128,56 @@ export const datatableVisualization: Visualization<
     ];
   },
 
-  renderLayerConfigPanel(domElement, props) {
-    const layer = props.state.layers.find(l => l.layerId === props.layerId);
-
-    if (layer) {
-      render(
-        <I18nProvider>
-          <DataTableLayer {...props} layer={layer} />
-        </I18nProvider>,
-        domElement
-      );
+  getLayerOptions({ state, frame, layerId }) {
+    const layer = state.layers.find(l => l.layerId === layerId);
+    if (!layer) {
+      return { dimensions: [] };
     }
+
+    const datasource = frame.datasourceLayers[layer.layerId];
+    const originalOrder = datasource.getTableSpec().map(({ columnId }) => columnId);
+    // When we add a column it could be empty, and therefore have no order
+    const sortedColumns = Array.from(new Set(originalOrder.concat(layer.columns)));
+
+    return {
+      dimensions: [
+        {
+          layerId: state.layers[0].layerId,
+          accessors: sortedColumns,
+          dimensionId: 'columns',
+          dimensionLabel: i18n.translate('xpack.lens.datatable.columns', {
+            defaultMessage: 'Columns',
+          }),
+          supportsMoreColumns: true,
+          filterOperations: () => true,
+        },
+      ],
+    };
+  },
+
+  setDimension({ layerId, columnId, prevState }) {
```

Comment:
nit: for readability when you compare with `removeDimension` params order
```suggestion
  setDimension({ prevState, columnId, layerId }) {
```

Reply:
Done